### ISSUE TITLE
[FEATURE] Implement LSTM Pitch Sequence Model (Optional Ensemble Component)

### ISSUE THREAD

**Issue by cbwinslow** [cbwinslow/retrosheet#156]
### 🎯 **Objective**

Develop a Long Short-Term Memory (LSTM) neural network model as **one component** of a multi-model ensemble system for predicting the next pitch in a baseball sequence. This model will coexist with other model types (XGBoost, Markov Chains, Transformers) and contribute to ensemble predictions through voting/weighting mechanisms.

### 📋 **Issue Details**
- **Type**: Feature (Optional Component)
- **Priority**: Medium (one of many model options)
- **Status**: Planning
- **Assignee**: Agent Cascade
- **Parent Issue**: #153 (EPIC: Enhanced Multi-Model ML Architecture)
- **Dependencies**: Phase 5 Unified Scheduler (COMPLETED)
- **Target Completion**: May 10, 2026

### 🧠 **LSTM Model Role in Ensemble**

The LSTM model will serve as **one approach** among multiple prediction models:

#### **Model Ecosystem Context**
```python
# LSTM is one of many models in the ensemble
class ModelRegistry:
    def __init__(self):
        self.available_models = {
            'lstm_sequence': LSTMModel(),           # This implementation
            'xgboost_hierarchical': XGBoostModel(),  # Tree-based approach
            'markov_chain': MarkovModel(),           # Probabilistic approach
            'transformer_attention': TransformerModel(), # Attention-based
            'bayesian_inference': BayesianModel(),   # Probabilistic inference
            'random_forest': RandomForestModel(),    # Ensemble trees
        }
```

#### **When LSTM Excels**
- **Sequential Patterns**: Long pitch sequences with temporal dependencies
- **Context Memory**: Remembering patterns from earlier in at-bat
- **Complex Relationships**: Non-linear pitch sequencing patterns
- **Player-Specific Patterns**: Individual pitcher tendencies over time

#### **When Other Models May Be Better**
- **Simple Count States**: Markov chains excel at basic count transitions
- **Tabular Features**: XGBoost better for structured feature data
- **Quick Predictions**: Markov models faster for simple cases
- **Interpretability**: Tree-based models more explainable

#### **Core LSTM Model**
```python
class LSTMModel(BaseModel):
    """LSTM model for pitch sequence prediction - one ensemble component"""
    
    def __init__(self, config: LSTMConfig):
        super().__init__("lstm_sequence")
        self.config = config
        
        # Embedding layers
        self.pitch_embedding = nn.Embedding(config.vocab_size, config.pitch_embed_dim)
        self.count_embedding = nn.Embedding(config.count_vocab_size, config.count_embed_dim)
        
        # LSTM layers
        self.lstm = nn.LSTM(
            input_size=config.input_dim,
            hidden_size=config.hidden_dim,
            num_layers=config.num_layers,
            dropout=config.dropout,
            batch_first=True
        )
        
        # Output heads (multi-task learning)
        self.pitch_type_head = nn.Linear(config.hidden_dim, config.vocab_size)
        self.swing_decision_head = nn.Linear(config.hidden_dim, 1)
        self.location_head = nn.Linear(config.hidden_dim, 2)
        
        # Confidence estimation
        self.confidence_head = nn.Linear(config.hidden_dim, 1)
        
    def predict(self, features: PredictionFeatures) -> ModelPrediction:
        """Standard interface for ensemble integration"""
        with torch.no_grad():
            outputs = self.forward(
                features.pitch_sequence,
                features.count_sequence,
                features.context_features
            )
            
            return ModelPrediction(
                model_name=self.name,
                pitch_probabilities=torch.softmax(outputs['pitch_type'], dim=-1),
                swing_probability=outputs['swing_decision'],
                location_prediction=outputs['location'],
                confidence_score=torch.sigmoid(outputs['confidence']),
                metadata={'model_type': 'lstm', 'sequence_length': len(features.pitch_sequence)}
            )
```

#### **Configuration**
```python
@dataclass
class LSTMConfig:
    # Model architecture
    vocab_size: int = 27  # Retrosheet pitch symbols
    count_vocab_size: int = 13  # Count states (0-0 to 3-2)
    pitch_embed_dim: int = 128
    count_embed_dim: int = 32
    hidden_dim: int = 256
    num_layers: int = 3
    dropout: float = 0.3
    
    # Training parameters
    learning_rate: float = 0.001
    batch_size: int = 64
    epochs: int = 50
    early_stopping_patience: int = 10
    
    # Sequence parameters
    max_sequence_length: int = 10  # Last 10 pitches
    min_sequence_length: int = 3
    
    # Ensemble integration
    confidence_threshold: float = 0.6  # Minimum confidence to contribute
    weight_in_ensemble: float = 1.0   # Relative weight in voting
```

#### **Training Data Preparation**
```python
class LSTMDataset(Dataset):
    """Dataset specific to LSTM model training"""
    
    def __init__(self, seasons: List[int], config: LSTMConfig):
        self.config = config
        self.data = self._load_sequence_data(seasons)
        
    def _load_sequence_data(self, seasons: List[int]) -> List[Dict]:
        """Load data optimized for LSTM sequential learning"""
        query = f"""
        SELECT 
            game_id,
            plate_appearance_id,
            pitch_index,
            count_label,
            raw_symbol,
            next_pitch_symbol,
            batter_id,
            pitcher_id,
            pre_pitch_balls,
            pre_pitch_strikes,
            inning,
            outs,
            balls,
            strikes,
            score_diff,
            -- Additional context for LSTM
            LAG(raw_symbol, 1) OVER (PARTITION BY plate_appearance_id ORDER BY pitch_index) as prev_pitch,
            LAG(raw_symbol, 2) OVER (PARTITION BY plate_appearance_id ORDER BY pitch_index) as prev_2_pitch,
            COUNT(*) OVER (PARTITION BY plate_appearance_id ORDER BY pitch_index ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) as pitch_number_in_pa
        FROM pitch_sequence.training_rows
        WHERE season IN ({','.join(map(str, seasons))})
        AND next_pitch_symbol IS NOT NULL
        ORDER BY game_id, plate_appearance_id, pitch_index
        """
        
        return self._create_lstm_sequences(query)
    
    def _create_lstm_sequences(self, raw_data: List[Dict]) -> List[Dict]:
        """Create sequences optimized for LSTM learning"""
        sequences = []
        
        for pa_id, pa_pitches in self._group_by_pa(raw_data):
            # Create overlapping sequences for LSTM
            for i in range(len(pa_pitches) - self.config.min_sequence_length):
                # Variable length sequences (3 to 10 pitches)
                seq_start = max(0, i - self.config.max_sequence_length + 1)
                seq_end = i + 1
                
                sequence = pa_pitches[seq_start:seq_end]
                target_pitch = pa_pitches[i]['next_pitch_symbol']
                
                if len(sequence) >= self.config.min_sequence_length:
                    sequences.append({
                        'sequence': sequence,
                        'target': target_pitch,
                        'context': self._extract_lstm_context(pa_pitches[i]),
                        'sequence_length': len(sequence)
                    })
        
        return sequences
```

#### **LSTM-Specific Training**
```python
class LSTMTrainer:
    """Training pipeline specific to LSTM model"""
    
    def __init__(self, config: LSTMConfig):
        self.config = config
        self.model = LSTMModel(config)
        self.optimizer = torch.optim.Adam(self.model.parameters(), lr=config.learning_rate)
        
        # Multi-task loss weighting
        self.pitch_loss_fn = nn.CrossEntropyLoss()
        self.swing_loss_fn = nn.BCELoss()
        self.location_loss_fn = nn.MSELoss()
        self.confidence_loss_fn = nn.BCELoss()
        
    def train(self, train_data: LSTMDataset, val_data: LSTMDataset):
        """Train LSTM with ensemble integration focus"""
        best_val_loss = float('inf')
        patience_counter = 0
        
        for epoch in range(self.config.epochs):
            # Training phase
            train_loss = self._train_epoch(train_data)
            
            # Validation phase
            val_loss, val_metrics = self._validate_epoch(val_data)
            
            # Ensemble-specific metrics
            ensemble_score = self._calculate_ensemble_score(val_metrics)
            
            # Early stopping based on ensemble contribution
            if val_loss < best_val_loss and ensemble_score > 0.7:
                best_val_loss = val_loss
                patience_counter = 0
                self._save_checkpoint(f'lstm_best_ensemble_{epoch}.pth')
            else:
                patience_counter += 1
                if patience_counter >= self.config.early_stopping_patience:
                    break
            
            print(f"Epoch {epoch}: Loss={train_loss:.4f}, Ensemble_Score={ensemble_score:.3f}")
    
    def _calculate_ensemble_score(self, metrics: Dict[str, float]) -> float:
        """Calculate how well this model contributes to ensemble"""
        # Weight accuracy, confidence, and diversity
        accuracy_weight = 0.5
        confidence_weight = 0.3
        diversity_weight = 0.2
        
        accuracy_score = metrics.get('pitch_accuracy', 0.0)
        confidence_score = metrics.get('avg_confidence', 0.0)
        diversity_score = metrics.get('prediction_diversity', 0.0)
        
        return (accuracy_weight * accuracy_score + 
                confidence_weight * confidence_score + 
                diversity_weight * diversity_score)
```

#### **LSTM-Specific Metrics**
- **Pitch Type Accuracy**: >70% (contributes to ensemble >75%)
- **Swing Decision Accuracy**: >65% (ensemble contribution)
- **Location Prediction**: MAE <1 foot
- **Confidence Calibration**: Brier score <0.2
- **Sequence Learning**: >80% accuracy on sequences >5 pitches

#### **Ensemble Integration**
- **Confidence Threshold**: >0.6 to contribute to ensemble
- **Diversity Score**: >0.3 (different predictions from other models)
- **Weight Contribution**: 0.8-1.2 relative weight in ensemble
- **Fallback Performance**: >60% accuracy when other models fail

#### **Training Performance**
- **Training Time**: <2 hours for full dataset
- **Inference Latency**: <50ms per prediction
- **Memory Usage**: <1GB for model and data
- **Convergence**: <20 epochs to reach ensemble contribution threshold

#### **LSTM-Specific Testing**
```python
class LSTMValidator:
    """Validation for LSTM model as ensemble component"""
    
    def validate_lstm(self, test_data: LSTMDataset) -> Dict[str, float]:
        """Comprehensive LSTM validation"""
        metrics = {
            'pitch_accuracy': 0.0,
            'swing_accuracy': 0.0,
            'location_mae': 0.0,
            'confidence_calibration': 0.0,
            'sequence_learning': 0.0,
            'ensemble_diversity': 0.0
        }
        
        # Test sequence length performance
        for seq_len in [3, 5, 7, 10]:
            seq_metrics = self._test_sequence_length(test_data, seq_len)
            metrics[f'seq_{seq_len}_accuracy'] = seq_metrics['accuracy']
        
        # Test ensemble contribution
        ensemble_metrics = self._test_ensemble_contribution(test_data)
        metrics.update(ensemble_metrics)
        
        return metrics
    
    def _test_ensemble_contribution(self, test_data: LSTMDataset) -> Dict[str, float]:
        """Test how LSTM contributes to ensemble"""
        # Compare LSTM predictions to other model types
        lstm_predictions = self._get_lstm_predictions(test_data)
        other_predictions = self._get_other_model_predictions(test_data)
        
        # Calculate diversity and complementarity
        diversity_score = self._calculate_prediction_diversity(lstm_predictions, other_predictions)
        complementarity = self._calculate_complementarity(lstm_predictions, other_predictions)
        
        return {
            'ensemble_diversity': diversity_score,
            'ensemble_complementarity': complementarity,
            'unique_coverage': self._calculate_unique_coverage(lstm_predictions, other_predictions)
        }
```

#### **Ensemble Registration**
```python
# Register LSTM as one model in the ensemble
def register_lstm_model(ensemble: MultiModelEnsemble):
    """Register LSTM model in the multi-model system"""
    
    lstm_config = LSTMConfig(
        vocab_size=27,
        hidden_dim=256,
        confidence_threshold=0.6,
        weight_in_ensemble=1.0
    )
    
    lstm_model = LSTMModel(lstm_config)
    
    # Model metadata for ensemble selection
    lstm_metadata = ModelMetadata(
        name="lstm_sequence",
        model_type="sequential_deep_learning",
        strengths=["long_sequences", "temporal_patterns", "player_specific"],
        weaknesses=["simple_count_states", "interpretability", "speed"],
        suitable_contexts=["long_at_bats", "complex_sequences", "player_tendencies"],
        confidence_threshold=0.6,
        expected_latency_ms=45,
        memory_footprint_mb=800
    )
    
    ensemble.model_registry.register_model("lstm_sequence", lstm_model, lstm_metadata)
```

#### **Context-Aware Selection**
```python
class ModelSelector:
    """Select models based on context - LSTM gets selected for appropriate situations"""
    
    def select_models(self, context: PredictionContext) -> List[str]:
        """Select models suitable for current context"""
        selected_models = []
        
        # LSTM excels in these situations
        if self._should_use_lstm(context):
            selected_models.append("lstm_sequence")
        
        # Add other models based on context
        if context.sequence_length >= 5:
            selected_models.append("lstm_sequence")
        if context.pitch_count_in_pa >= 3:
            selected_models.append("lstm_sequence")
        if context.has_complex_patterns:
            selected_models.append("lstm_sequence")
        
        return selected_models
    
    def _should_use_lstm(self, context: PredictionContext) -> bool:
        """Determine if LSTM is appropriate for this context"""
        return (
            context.sequence_length >= 3 and
            context.has_sufficient_history and
            not context.needs_fast_prediction and
            context.pitch_count_in_pa <= 8  # LSTM better for mid-PA
        )
```

#### **Model Performance**
- [ ] **Pitch Type Accuracy**: >70% on test set
- [ ] **Sequence Learning**: >80% on sequences >5 pitches
- [ ] **Confidence Calibration**: Brier score <0.2
- [ ] **Inference Latency**: <50ms per prediction

#### **Ensemble Integration**
- [ ] **Ensemble Contribution**: Positive impact on ensemble accuracy
- [ ] **Diversity Score**: >0.3 (different from other models)
- [ ] **Context Selection**: Selected in appropriate situations
- [ ] **Weight Balance**: Contributes appropriately to ensemble voting

#### **Code Quality**
- [ ] **Test Coverage**: >90% test coverage
- [ ] **Ensemble Interface**: Implements BaseModel interface correctly
- [ ] **Documentation**: Complete integration documentation
- [ ] **Performance**: Optimized for ensemble use

### 🎯 **Next Steps**

1. **Implement LSTM model**: Build core architecture with ensemble interface
2. **Train on sequence data**: Focus on sequential learning patterns
3. **Validate ensemble contribution**: Test diversity and complementarity
4. **Register in ensemble**: Add to multi-model system
5. **Test context selection**: Verify appropriate model selection
6. **Monitor performance**: Track ensemble contribution over time

### 📝 **Acceptance Criteria**

- [ ] LSTM model implements BaseModel interface for ensemble integration
- [ ] Model trained on sequential pitch data with >70% accuracy
- [ ] Demonstrates positive contribution to ensemble accuracy
- [ ] Selected appropriately by context-aware model selector
- [ ] Provides diverse predictions from other model types
- [ ] Meets performance targets for ensemble use
- [ ] Complete integration testing with multi-model system

---

**Status**: 📋 **PLANNING**

**Target Start**: May 6, 2026

**Target Completion**: May 10, 2026

**Dependencies**: Phase 5 Unified Scheduler (COMPLETED)

**Blocked By**: None

**Assignee**: Agent Cascade

**Note**: This is **one component** of a multi-model ensemble system, not the primary prediction approach.

**Comment by cbwinslow** [comment]
✅ **COMPLETED** - LSTM Pitch Sequence Model fully implemented.

Implementation includes:
- baseball/models/lstm/sequential_model.py - 402 lines of comprehensive LSTM model
- baseball/models/lstm/pitch_encoder.py - Pitch encoding for sequence processing
- baseball/models/lstm/attention.py - 8K lines of attention mechanism
- baseball/models/lstm/__init__.py - Complete module exports
- Comprehensive test suite (tests/models/test_lstm_sequence.py) with 400+ lines

Features implemented:
- Sequential LSTM with configurable architecture
- Attention mechanisms for sequence processing
- Pitch encoder for vocabulary management
- Variable sequence length support
- GPU acceleration support
- Model serialization and loading
- Batch prediction capabilities
- Performance metrics and validation
- Integration with ensemble system

All LSTM sequence model components fully implemented with comprehensive functionality and testing coverage. Ready for production use as optional ensemble component.
